## Precision

To derive the correlation matrices I am assuming a discrete time series of random values. I expect that the derivation in continuous time will proceed analogously.

Assume $x$ is a stationary stochastic process. Then we can define:

\begin{align*}
E(\bar{x}) = \mu &= \frac{1}{n} \sum_{i=1}^{n}{x_i} \\
V(x) = \sigma^2 = 1/\pi &= \frac{1}{n} \sum_{i=1}^{n-h}{(x_i - \bar{x})(x_i -\bar{x})} \\
C(x_n,x_{n+h}) = \nu(h) &= \frac{1}{n} \sum_{i=1}^{n-h}{(x_i - \bar{x})(x_{i+h} - \bar{x})} \\
\rho(h) = \nu(h)/\sigma^2 &= \pi \nu(h) \\
\end{align*}

where $\mu$ is the mean, $\sigma^2$ is the variance, $\pi$ is the precision, $\nu(h)$ is the cross-covariance, and $\rho(h)$ is the autocorrelation function, with $\rho(0) = 1$ by definition, when $n \to \infty$.

Without loss of generality, we can assume $x$ is a zero mean process so we can simplify to:

\begin{align*}
\mu &= 0 \\
\sigma^2 = 1/\pi &= \frac{1}{n} \sum_{i=1}^{n-h}{x_i x_i} \\
\nu(h) &= \frac{1}{n} \sum_{i=1}^{n-h}{x_i x_{i+h}} \\
\rho(h) &= \pi \nu(h) \\
\end{align*}

If $x$ is not a single variable, but a set of variables, then the cross-covariance and precision of the process are matrices:

\begin{align*}
\Sigma^2 = \Pi^{-1} = \frac{1}{n} \sum_{i=1}^{n-h}{x_i x_i^T} \\
\end{align*}

If $\tilde x$ is a vector of generalised coordinates, we have:

\begin{align*}
\tilde x &= [x, \dot x, \ddot x, \dddot x, \cdots] \\
\Sigma^2 &= \frac{1}{n} \sum_{i=1}^{n-h}{\tilde x_i \tilde x_i^T} \\
&= 
    \begin{bmatrix}
    C(x,x) & C(x,\dot{x}) & C(x,\ddot{x}) & \cdots \\
    C(\dot{x},x) & C(\dot{x},\dot{x}) & C(\dot{x},\ddot{x}) \\
    C(\ddot{x},x) & C(\ddot{x},\dot{x}) & C(\ddot{x},\ddot{x}) \\
    \vdots & & & \ddots
    \end{bmatrix}
\end{align*}

Because the levels in generalised coordinates are derivatives of the levels above, temporal correlations will exist between the levels. We can define the temporal derivative $x$:

\begin{align*}
\lim_{\Delta t \to 0} E \left( \dot{x}(t) - \frac{x(t + \Delta t) - x(t)}{\Delta t}\right) = 0 \\
\end{align*}

or, more informally:

\begin{align*}
\dot{x}(t) &= \frac{x(t + \Delta t) - x(t)}{\Delta t} ; \Delta t \to 0\\
\ddot{x}(t) &= \frac{\dot{x}(t + \Delta t) - \dot{x}(t)}{\Delta t} \\
&= \frac{x(t + \Delta t) - x(t)}{(\Delta t)^2} - \frac{x(t) - x(t - \Delta t)}{(\Delta t)^2}\\
&= \frac{x(t + \Delta t) - 2x(t) + x(t - \Delta t)}{(\Delta t)^2}; \Delta t \to 0\\
\end{align*}

Thus we can proceed as follows:

\begin{align*}
C(x,x) &= \sigma^2 \rho(0)\\
C(x,\dot{x}) &= \frac{1}{n} \sum_{i=1}^{n}{x_i \dot{x}_i} \\
&= \frac{1}{n} \sum_{i=1}^{n}{x(i) \frac{x(i + \Delta t) - x(i)}{\Delta t}} \\
&= \frac{1}{n} \sum_{i=1}^{n}{\frac{x(i) x(i + \Delta t)}{\Delta t} - \frac{x(i) x(i)}{\Delta t}} \\
&= \frac{\nu(\Delta t)}{\Delta t} - \frac{\nu(0)}{\Delta t} \\
&= \frac{\nu(\Delta t) - \nu(0)}{\Delta t} \\
&= \sigma^2 \frac{\rho(\Delta t) - \rho(0)}{\Delta t} \\
&= \sigma^2 \dot{\rho}(0)\\
C(x,\ddot{x}) &= \frac{1}{n} \sum_{i=1}^{n}{x(i) \frac{x(i + \Delta t) - 2x(i) + x(i - \Delta t)}{(\Delta t)^2}} \\
&= \frac{1}{n} \sum_{i=1}^{n}{\frac{x(i) x(i + \Delta t) - 2x(i) x(i) + x(i) x(i - \Delta t)}{(\Delta t)^2}} \\
&= \frac{\nu(\Delta t) - 2\nu(0) + \nu(- \Delta t)}{(\Delta t)^2} \\
&= \sigma^2 \frac{\dot{\rho}(\Delta t) - \dot{\rho}(0)}{\Delta t} \\
&= \sigma^2 \ddot{\rho}(0) \\
C(\dot{x},\dot{x}) &= \frac{1}{n} \sum_{i=1}^{n}{\dot{x}_i \dot{x}_i} \\
&= \frac{1}{n} \sum_{i=1}^{n}{\frac{x(i + \Delta t) - x(i)}{\Delta t} \frac{x(i + \Delta t) - x(i)}{\Delta t}} \\
&= \frac{1}{n} \sum_{i=1}^{n}{\frac{x(i + \Delta t) x(i + \Delta t) - 2x(i) x(i + \Delta t) + x(i)x(i)}{(\Delta t)^2}} \\
&= \frac{2\nu(0) - 2\nu(\Delta t)}{(\Delta t)^2} \\
&= \sigma^2 \frac{2\rho(0) - 2\rho(\Delta t)}{(\Delta t)^2} \\
&= -\sigma^2 \frac{\rho(\Delta t) - 2\rho(0) + \rho(-\Delta t)}{(\Delta t)^2} \;(\text{since}\; \rho(i) = \rho(-i))\\
&= -\sigma^2 \frac{\frac{\rho(\Delta t) - \rho(0)}{\Delta t} - \frac{\rho(0) - \rho(-\Delta t)}{\Delta t}}{\Delta t} \\
&= -\sigma^2 \frac{\dot{\rho}(\Delta t) - \dot{\rho}(0)}{\Delta t} \\
&= -\sigma^2 \ddot{\rho}(0)\\
C(\dot{x},\ddot{x}) &= \frac{1}{n} \sum_{i=1}^{n}{\dot{x}_i \ddot{x}_i} \\
&= \frac{1}{n} \sum_{i=1}^{n}{\dot{x}(i) \frac{\dot{x}(i + \Delta t) - \dot{x}(i)}{\Delta t}} \\
&= -\sigma^2 \frac{\ddot{\rho}(\Delta t) - \ddot{\rho}(0)}{\Delta t} \\
&= -\sigma^2 \dddot{\rho}(0)\\
C(\ddot{x},\ddot{x}) &= \frac{1}{n} \sum_{i=1}^{n}{\ddot{x}_i \ddot{x}_i} \\
&= \frac{1}{n} \sum_{i=1}^{n}{\frac{\dot{x}(i + \Delta t) - \dot{x}(i)}{\Delta t} \frac{\dot{x}(i + \Delta t) - \dot{x}(i)}{\Delta t}} \\
&= -\sigma^2 \frac{2\ddot{\rho}(0) - 2\ddot{\rho}(\Delta t)}{(\Delta t)^2} \\
&= \sigma^2 \frac{\ddot{\rho}(\Delta t) - 2\ddot{\rho}(0) + \ddot{\rho}(\Delta t)}{(\Delta t)^2} \;(\text{since}\; \ddot{\rho}(i) = \ddot{\rho}(-i))\\
&= \sigma^2 \ddot{\ddot{\rho}}(0)
\end{align*}

We can proceed similarly for higher orders of derivatives to get:

\begin{align*}
\Sigma^2 = \sigma^2 S^{-1} = \sigma^2
    \begin{bmatrix}
    \rho(0) & \dot{\rho}(0) & \ddot{\rho}(0) & -\dddot{\rho}(0) & \cdots \\
    \dot{\rho}(0) & -\ddot{\rho}(0) & -\dddot{\rho}(0) & -\ddot{\ddot{\rho}}(0)\\
    \ddot{\rho}(0) & -\dddot{\rho}(0) & \ddot{\ddot{\rho}}(0) & \ddot{\dddot{\rho}}(0) \\
    -\dddot{\rho}(0) & -\ddot{\ddot{\rho}}(0) & \ddot{\dddot{\rho}}(0) & -\dddot{\dddot{\rho}}(0) \\
    \vdots & & & & \ddots
    \end{bmatrix}
\end{align*}

or

\begin{align*}
\tilde \pi &= {\Sigma^2}^{-1} = \pi S
\end{align*}

Now if $\tilde x$ is not only the vector of generalised coordinates of a single random variable, but a set of variables, the correlations between the levels of the generalised coordinates stay the same, but the scalar precision $\pi$ is replaced by a precision matrix $\Pi$, and the full precision matrix is the Kronecker tensor product $\tilde \Pi = \Pi \otimes S$.

## Autocorrelation function

The autocorrelation function $\rho(h)$ for a given delay $h$ is symmetric around $h=0$. In other words, $\rho$ is an even function. This implies directly that $\dot{\rho}(h)$ is odd, $\ddot{\rho}(h)$ is even, etc. Odd functions evaluated at $0$ yield $0$. Also $\rho(0) = 1$ by construction. Thus we can simplyfy $S$ as:

\begin{align*}
S =
    \begin{bmatrix}
    1 & 0 & \ddot{\rho}(0) & 0 & \cdots \\
    0 & -\ddot{\rho}(0) & 0 & -\ddot{\ddot{\rho}}(0)\\
    \ddot{\rho}(0) & 0 & \ddot{\ddot{\rho}}(0) & 0 \\
    0 & -\ddot{\ddot{\rho}}(0) & 0 & -\dddot{\dddot{\rho}}(0) \\
    \vdots & & & & \ddots
    \end{bmatrix}^{-1}
\end{align*}

$S$ can be evaluated for any analytic autocorrelation function. If we assume, for convenience, that the temporal correlation of all innovations have the same Gaussian form with precision parameter $\gamma$, we can proceed with:

\begin{align*}
\rho(h) &= \exp{-\frac{\gamma}{2} h^2} \\
\dot{\rho}(h) &= -\gamma h \rho(h) \\
\ddot{\rho}(h) &= -\gamma \rho(h) + (\gamma h)^2 \rho(h) \\
\dddot{\rho}(h) &= \gamma^2 h \rho(h) + 2 \gamma^2 h \rho(h) - (\gamma h)^3 \rho(h) \\
&= 3 \gamma^2 h \rho(h) - (\gamma h)^3 \rho(h) \\
\ddot{\ddot{\rho}}(h) &= 3 \gamma^2 \rho(h) - 3 \gamma^3 h^2 \rho(h) - 3 \gamma^3 h^2 \rho(h) + (\gamma h)^4 \rho(h) \\
&= 3 \gamma^2 \rho(h) - 6 \gamma^3 h^2 \rho(h) + (\gamma h)^4 \rho(h) \\
\ddot{\dddot{\rho}}(h) &= - 3 \gamma^3 h \rho(h) - 12 \gamma^3 h \rho(h) + 6 \gamma^4 h^3 \rho(h) + 4 \gamma^4 h^3 \rho(h) - (\gamma h)^5 \rho(h) \\
&= - 15 \gamma^3 h \rho(h) + 10 \gamma^4 h^3 \rho(h) - (\gamma h)^5 \rho(h) \\
\dddot{\dddot{\rho}}(h) &= -15 \gamma^3 \rho(h) + 15 \gamma^4 h^2 \rho(h) + 30 \gamma^4 h^2 \rho(h) - 10 \gamma^5 h^4 \rho(h) - 5 \gamma^5 h^4 \rho(h) + (\gamma h)^6 \rho(h) \\
&= -15 \gamma^3 \rho(h) + 45 \gamma^4 h^2 \rho(h) - 15 \gamma^5 h^4 \rho(h) + (\gamma h)^6 \rho(h) \\
\end{align*}

To calculate the higher order derivatives, the Hermite polynomials can be used as a shortcut to find the relevant coefficients.

Evaluating the above at $h = 0$ yields:

\begin{align*}
\rho(0) &= 1 \\
\dot{\rho}(0) &= 0 \\
\ddot{\rho}(0) &= -\gamma \\
\dddot{\rho}(0) &= 0 \\
\ddot{\ddot{\rho}}(0) &= 3 \gamma^2 \\
\ddot{\dddot{\rho}}(0) &= 0 \\
\dddot{\dddot{\rho}}(0) &= -15 \gamma^3 \\
\end{align*}

From the Hermite polynomials we can quickly find the coefficient for all even order derivatives of order $n$:

\begin{align*}
c_n = n! \frac{(-1)^{n/2}}{2^{n/2+1}(n/2)!}
\end{align*}



In [7]:
from math import factorial as fac
print("Derivatives of 𝜌")
for m in range(6):
    print("{:<3} {:>8} γ^{:}".format(2*m, (-1)**m * fac(2*m) / fac(m) / 2**(m), m))
    print("{:<3} {:>8}".format(2*m + 1, 0))

Derivatives of 𝜌
0        1.0 γ^0
1          0
2       -1.0 γ^1
3          0
4        3.0 γ^2
5          0
6      -15.0 γ^3
7          0
8      105.0 γ^4
9          0
10    -945.0 γ^5
11         0


\begin{align*}
S &=
    \begin{bmatrix}
    1 & 0 & \ddot{\rho}(0) & 0 & \cdots \\
    0 & -\ddot{\rho}(0) & 0 & -\ddot{\ddot{\rho}}(0)\\
    \ddot{\rho}(0) & 0 & \ddot{\ddot{\rho}}(0) & 0 \\
    0 & -\ddot{\ddot{\rho}}(0) & 0 & -\dddot{\dddot{\rho}}(0) \\
    \vdots & & & & \ddots
    \end{bmatrix}^{-1} \\
&=
    \begin{bmatrix}
    1 & 0 & -\gamma & 0 & 3 \gamma^2 & 0 \\
    0 & \gamma & 0 & -3 \gamma^2 & 0 & 15 \gamma^3 \\
    -\gamma & 0 & 3 \gamma^2 & 0 & 15 \gamma^3 & 0 \\
    0 & -3 \gamma^2 & 0 & 15 \gamma^3 & 0 & 105 \gamma^4 \\
    3 \gamma^2 & 0 & 15 \gamma^3 & 0 & 105 \gamma^4 & 0 \\
    0 & 15 \gamma^3 & 0 & 105 \gamma^4 & 0 & 945 \gamma^5 \\
    \end{bmatrix}^{-1}
\end{align*}

Note, to get Friston's version, we need to replace $\gamma$ with $\gamma / 2$. I'm not sure where this discrepancy comes from. It could just be a matter of how you define precision, but in that case his version would be odd, as it results in $\rho(h) = \exp(-\gamma h^2/4)$. 